Convert max effects to normal move effects

AlexOn1ine opened the pull request:
Converted Max Move Effects to normal Move Effects so they can be used for all moves.

AlexOn1ine commented:
Ready for review. 

Note: the function for move effects is getting bigger and bigger so I think it would make sense to turn it into a function table but should be a separate PR

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -4304,6 +4301,304 @@ void SetMoveEffect(bool32 primary, bool32 certain)
+                    gBattlescriptCurrInstr = BattleScript_EffectRaiseStatAllies;
+                }
+                break;
+            case MOVE_EFFECT_LOWER_ATTACK:
+            case MOVE_EFFECT_LOWER_DEFENSE:
+            case MOVE_EFFECT_LOWER_SPEED:
+            case MOVE_EFFECT_LOWER_SP_ATK:
+            case MOVE_EFFECT_LOWER_SP_DEF:
```
All of these effects affect both opponents, so they should at least have the `_FOES` suffix

AlexOn1ine commented on include/constants/battle.h:
```diff
@@ -322,102 +322,152 @@ enum BattleWeather
-#define MOVE_EFFECT_TERA_BLAST          78
-#define MOVE_EFFECT_ORDER_UP            79
-#define MOVE_EFFECT_ION_DELUGE          80
-#define MOVE_EFFECT_AROMATHERAPY        81 // No functionality yet
```
can I do it in a follow up pr? I don't want to write a test right now.

AsparagusEduardo replied:
alright

AlexOn1ine commented on src/battle_ai_util.c:
```diff
@@ -2398,8 +2398,8 @@ static inline bool32 IsMoveSleepClauseTrigger(u32 move)
-        case MAX_EFFECT_EFFECT_SPORE_FOES:
-        case MAX_EFFECT_YAWN_FOE:
+        case MOVE_EFFECT_EFFECT_SPORE_FOES:
+        case MOVE_EFFECT_YAWN_FOE:
```
I don't know. Idr but I think there was a reason why it wasn't added here but we would need to ask Pawkkie. I would say this is out of scope in this pr though

Pawkkie replied:
Saw my name in the Discord channel lol

They are specifically omitted intentionally, that move effect is used for moves that have a secondary effect with a chance of applying sleep, which don't trigger sleep clause using Smogon's standard ruleset :)

ie. you can still use Relic Song if you've already slept another mon, etc.

Pawkkie replied:
I think that's a great idea, d'you want to do it here  or d'you want me to make a follow up PR later

AlexOn1ine replied:
follow up would be better I think

Pawkkie replied:
Sounds good, gonna make issue cuz will forget and can't rn :)

AlexOn1ine commented:
ready for re-review

AsparagusEduardo commented:
It's not compiling ^^;
I don't know I didn't think of this earlier, but probably `_SIDE` would be a better suffix

AlexOn1ine commented:
> It's not compiling ^^; I don't know I didn't think of this earlier, but probably `_SIDE` would be a better suffix

fixed compiling and renamed defines to `_SIDE`

AlexOn1ine commented:
Ready for re-review

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -4569,12 +4578,21 @@ void SetMoveEffect(bool32 primary, bool32 certain)
+                BattleScriptPush(gBattlescriptCurrInstr + 1);
+                gBattlescriptCurrInstr = BattleScript_EffectTormentSide;
+                break;
             case MOVE_EFFECT_MEAN_LOOK:
```
```suggestion
            case MOVE_EFFECT_PREVENT_ESCAPE_SIDE:
```

AsparagusEduardo commented on include/move.h:
```diff
@@ -122,7 +122,7 @@ struct MoveInfo
         u16 fixedDamage;
         u16 absorbPercentage;
-        u16 maxEffect;
+        u16 argMoveEffect;
```
This field should now be removed.

AsparagusEduardo commented on include/move.h:
```diff
@@ -489,9 +489,9 @@ static inline u32 GetMoveAbsorbPercentage(u32 moveId)
 }
 
-static inline u32 GetMoveMaxEffect(u32 moveId)
+static inline u32 GetArgumentMoveEffect(u32 moveId)
```
Same with this function

AsparagusEduardo commented on src/battle_dynamax.c:
```diff
@@ -322,7 +322,7 @@ u8 GetMaxMovePower(u32 move)
     u8 tier;
     // G-Max Drum Solo, G-Max Hydrosnipe, and G-Max Fireball always have 160 base power.
-    if (GetMoveMaxEffect(GetMaxMove(gBattlerAttacker, move)) == MAX_EFFECT_FIXED_POWER)
+    if (GetArgumentMoveEffect(GetMaxMove(gBattlerAttacker, move)) == MOVE_EFFECT_FIXED_POWER)
```
Somehow the test `"(DYNAMAX) G-Max Hydrosnipe has fixed power and ignores abilities"` didn't fail despite removing `MOVE_EFFECT_FIXED_POWER` from `argMoveEffect`. This should be corrected

AlexOn1ine replied:
Yeah, there seems to be a problem

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -4304,6 +4301,322 @@ void SetMoveEffect(bool32 primary, bool32 certain)
+                BattleScriptPush(gBattlescriptCurrInstr + 1);
+                gBattlescriptCurrInstr = BattleScript_EffectHealOneSixthAllies;
+                break;
+            case MOVE_EFFECT_AROMATHERAPY:
```
```suggestion
            case MOVE_EFFECT_AROMATHERAPY_TEAM:
```

AlexOn1ine replied:
I saw your comment but I disagree with preemptively changing names. Right now Aromatherapy be default cures the team. By adding the team suffix you are implying that we can change it.

Applies to all other similar suggestions.

AsparagusEduardo replied:
Ah, you're right. I forgot Aromatherapy already did it for the whole team. Please disregard this comment.

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -4304,6 +4301,322 @@ void SetMoveEffect(bool32 primary, bool32 certain)
+                BattleScriptPush(gBattlescriptCurrInstr + 1);
+                gBattlescriptCurrInstr = BattleScript_EffectMeanLookSide;
+                break;
+            case MOVE_EFFECT_CRIT_PLUS:
```
```suggestion
            case MOVE_EFFECT_CRIT_PLUS_TEAM:
```

AlexOn1ine replied:
Shouldn't this be side instead of team?

AlexOn1ine replied:
renamed to side for now

AsparagusEduardo replied:
> renamed to side for now

Thank you

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -13554,7 +13867,7 @@ static void Cmd_tryspiteppreduce(void)
             s32 ppToDeduct = B_PP_REDUCED_BY_SPITE >= GEN_4 ? 4 : (Random() & 3) + 2;
             // G-Max Depletion only deducts 2 PP.
-            if (IsMaxMove(gCurrentMove) && GetMoveMaxEffect(gCurrentMove) == MAX_EFFECT_SPITE)
+            if (IsMaxMove(gCurrentMove) && GetArgumentMoveEffect(gCurrentMove) == MOVE_EFFECT_SPITE)
```
The test `"(DYNAMAX) G-Max Depletion takes away 2 PP from the target's last move"` didn't fail because it's not actually checking the amount of reduced PP. Based on this change, instead of reducing PP by 2, it will reduce 4 PP (when set `B_PP_REDUCED_BY_SPITE` is Gen4+ and randomly when set to Gen3 or lower, like regular Spite.

AlexOn1ine replied:
It didn't fail because it only tested for messages.

AsparagusEduardo replied:
> It didn't fail because it only tested for messages.

Yeah, that's what I meant. The test shouldn't pass the "takes away 2 PP" part because it reduces more.

I'll make a small PR to address this and stopping the PR from being blocked.

AsparagusEduardo commented on src/battle_anim_new.c:
```diff
@@ -9267,18 +9267,18 @@ void AnimTask_DynamaxGrowth(u8 taskId) // from CFRU
 void AnimTask_GetWeatherToSet(u8 taskId)
 {
-    switch (GetMoveMaxEffect(gCurrentMove))
+    switch (GetArgumentMoveEffect(gCurrentMove))
```
This function will do nothing because the move effect was removed from `argMoveEffect`.

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -4304,6 +4301,322 @@ void SetMoveEffect(bool32 primary, bool32 certain)
+                    gBattlescriptCurrInstr = BattleScript_EffectGravitySuccess;
+                }
+                break;
+            case MOVE_EFFECT_SANDBLAST_FOES:
+            case MOVE_EFFECT_FIRE_SPIN_FOES:
```
To keep consistency with the other renames
```suggestion
            case MOVE_EFFECT_SANDBLAST_SIDE:
            case MOVE_EFFECT_FIRE_SPIN_SIDE:
```

AsparagusEduardo commented:
In the future, I'd like to improve on secondary move effects by being able to specify if it they affect the whole side or not, so probably the names proposed here don't matter that much, but it'd be good for clarity.

Sorry if I'm being to nitpicky on this PR.

AlexOn1ine commented on src/battle_script_commands.c:
```diff
@@ -18379,62 +18425,59 @@ void BS_TrySetStatus1(void)
+    }
+}
+
+void BS_TrySetTorment(void)
```
we have, but both have different conditions. Some might only apply to dmax moves and others not? Maybe both have bugs

AlexOn1ine replied:
Can we just open an issue for that and handle it once we have vanilla testing data (or at least showdown)?

AsparagusEduardo replied:
opened a new issue

AlexOn1ine commented:
forgot to mention but ready for re-review
